### PR TITLE
Switch to using agentbeat versus all the other beats from the beats repository

### LINKED ISSUE #3364: Offer a lightweight Elastic Agent
When running Elastic Agent as a Daemon Set on Kubernetes it can soon become a rather heavy resource consumer within the cluster. It would be great if we offered a minimal image for situations like this and have Agent pull assets from an EPR. 
This is currently blocking one of my customers adopting Elastic agent. 

**Describe a specific use case for the enhancement or feature:**
If a customer with 100 nodes of Kubernetes want to monitor metrics from each node they are required to run Agent as a daemon set which contains all Beats when actually all they require is Metricbeat. This not only consumed memory but storage also. 

Ensure the Elastic Agent container can ship with the lightest of resources and pull and further assets when required. 

UPDATE: Changed title to widen the scope of this issue. VMs, Server, Edge devices and containerised systems should all be considered when looking in to this issue.

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Switches to used the `agentbeat` that comes from this PR: https://github.com/elastic/beats/pull/38183

## Why is it important?

Reduces the size of the shipped Elastic Agent by almost 50%

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- [x] I have added an integration test or an E2E test

## How to test this PR locally

```
$ EXTERNAL="true" SNAPSHOT="true" mage package
```

## Related issues

- Closes #3364 

## Logs

Component logs are even correct and still showing `metricbeat` or `filebeat` when the binary being executed is actual `agentbeat metricbeat` and `agentbeat filebeat` respectively.

### CODE REVIEW

**Comment by blakerouse** [comment]
Sonarqube is wrong all the changes in this PR are fully covered by unit tests, I have confirmed it with my own coverage report review. I assume its counting the magefile contents in its percentage.

**Comment by blakerouse** [comment]
CI is broken so I am unable to get a green CI run on this PR, but I do see in the CI jobs that the elastic-agent was built with agentbeat. That is a great sign that it was built with it and was wanting to test elastic-agent with the agentbeat.

**Review by cmacknz — APPROVED** [review]

**Comment by cmacknz** [comment]
LGTM pending the tests passing

**Comment by blakerouse** [comment]
@cmacknz I had to make some changes to allow the elastic-agent-cloud image to be built. I have tested that it works and allows filebeat and metricbeat to be executed from there original paths all transparently running agentbeat.

https://github.com/elastic/elastic-agent/pull/4516/files#diff-9e8bcd1359f1e06bbdf2c89db711c6c400ed4142cf9764ff03e6e963013122e5R284

Probably would be good to get someone from cloud to review this change, I don't know who to ask.

**Comment by cmacknz** [comment]
Best thing to do for cloud is to test it by creating a deployment from your branch. https://github.com/elastic/elastic-agent?tab=readme-ov-file#testing-on-elastic-cloud

Make sure this branch has https://github.com/elastic/elastic-agent/commit/eca5bc714b9a5569e733644641deabbf4b949e3b or it will fail for unrelated reasons. There is also possibly a fleet-server issue you may see that is unrelated, that won't be fixed until the agentbeat CI fix is merged.

File: dev-tools/packaging/packages.yml
Comment (pchila) [inline dev-tools/packaging/packages.yml:284]: Why were the two files moved out from the generic docker spec? Are these only used in `-cloud` images ?

File: magefile.go
Comment (pchila) [inline magefile.go:1020]: Since we can specify a BEAT_VERSION env var, shouldn't we make a conditional to fetch agentbeat only from 8.14.0 onwards ? If we don't do that packaging will fail for any beat version lower than 8.14.0

Comment (blakerouse) [reply magefile.go:1020]: Each version of elastic-agent is branched so this change will take effect 8.14+ which is where it applies.

File: dev-tools/mage/manifest/manifest.go
Comment (pchila) [inline dev-tools/mage/manifest/manifest.go:89]: (Similar issue as the one above) This bit of code can be used with manifests older than 8.14 (8.13.x for example) and the repackaging can happen on any branch since it's independent from the actual code...

This change can break if a repackage for 8.13 is launched using the magefile from main....
Did you check with the early-agent-releases what this change implies for them ? cc @pierrehilbert @cmacknz

Comment (pchila) [reply dev-tools/mage/manifest/manifest.go:89]: > > on which branch are we planning to call the mage downloadManifest target if we need to create an agent early release ?
> 
> Once we're fully ready, the Independent Agent Staging build will run the `elastic-agent-package` pipeline on the "Stack Branch" for the release. So if we are building a `8.13.2+build202404171057` staging build, then it will run on the `8.13` branch for the `elastic-agent-package` pipeline.
> 
> We're going to start doing regular/daily builds starting soon on `8.13`, followed by `8.14` sometime later.

On 8.13 we don't have the new target to use elastic-agent-core artifacts that was introduced with #4403
The assumption I had when implementing that was that the pipeline could be triggered from anywhere (not true anymore after this change) and we already partially backported #4403 (for the env var rename) which will make the backport more complicated. I guess we need to sync on where you expect the `packageUsingDRA` mage target to exist

Comment (pchila) [reply dev-tools/mage/manifest/manifest.go:89]: Opened https://github.com/elastic/elastic-agent/pull/4589 to backport the repackaging logic on 8.13 branch

**Review by pchila — APPROVED** [review]

**Comment by cmacknz** [comment]
buildkite test it

**Comment by cmacknz** [comment]
The runtime leak tests are failing and might need some adjustment, it is trying to get the beat stats directly from each component, and now all the components are the same process.

https://github.com/elastic/elastic-agent/blob/main/testing/integration/agent_long_running_leak_test.go

**Comment by blakerouse** [comment]
Very strange. I would have expected that to work the same, as I tried to ensure it didn't adjust any naming or paths.

Trying to figure out what the issue is. Working on it.

**Comment by cmacknz** [comment]
I see this in the logs, not sure if it's the cause but it's going to break the monitoring data:

```json
"component":{"binary":"agentbeat","dataset":"elastic_agent.agentbeat","id":"log-default","type":"log"}
```

The dataset `elastic_agent.agentbeat` doesn't exist and Fleet won't create index permissions.

I can see several "cannot index event" errors I believe are related:

```json
{"log.level":"warn","@timestamp":"2024-04-17T17:46:53.872Z","message":"Cannot index event (status=403): dropping event! Enable debug logs to view the event and cause.","component":{"binary":"agentbeat","dataset":"elastic_agent.agentbeat","id":"filestream-monitoring","type":"filestream"},"log":{"source":"filestream-monitoring"},"ecs.version":"1.6.0","log.logger":"elasticsearch","log.origin":{"file.line":429,"file.name":"elasticsearch/client.go","function":"github.com/elastic/beats/v7/libbeat/outputs/elasticsearch.(*Client).bulkCollectPublishFails"},"service.name":"filebeat","ecs.version":"1.6.0"}
```

**Comment by blakerouse** [comment]
I am seeing something a little different:

```
{"log.level":"info","@timestamp":"2024-04-17T18:46:52.217-0400","message":"Metrics endpoint listening on: /tmp/elastic-agent/Hk6rvk9TDibMPcDvpl0jkLE-qDsHWVYL.sock (configured: unix:///tmp/elastic-agent/Hk6rvk9TDibMPcDvpl0jkLE-qDsHWVYL.sock)","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"log":{"source":"beat/metrics-monitoring"},"log.logger":"api","log.origin":{"file.line":71,"file.name":"api/server.go","function":"github.com/elastic/beats/v7/libbeat/api.(*Server).Start.func1"},"service.name":"metricbeat","ecs.version":"1.6.0","ecs.version":"1.6.0"}
```

**Comment by cmacknz** [comment]
Something is definitely up with the monitoring sockets. There are 4 unique PIDs but only 3 unix sockets:

```
❯ sudo elastic-agent status --output=full
┌─ fleet
│  └─ status: (STOPPED) Not enrolled into Fleet
└─ elastic-agent
   ├─ status: (HEALTHY) Running
   ├─ info
   │  ├─ id: 294ec647-1044-47ca-8c64-714f9b48dcbb
   │  ├─ version: 8.14.0
   │  └─ commit: d44fe9fac2a56d529ddf30833fd2b70535588249
   ├─ beat/metrics-monitoring
   │  ├─ status: (HEALTHY) Healthy: communicating with pid '75187'
   │  ├─ beat/metrics-monitoring
   │  │  ├─ status: (HEALTHY) Healthy
   │  │  └─ type: OUTPUT
   │  └─ beat/metrics-monitoring-metrics-monitoring-beats
   │     ├─ status: (HEALTHY) Healthy
   │     └─ type: INPUT
   ├─ filestream-monitoring
   │  ├─ status: (HEALTHY) Healthy: communicating with pid '75186'
   │  ├─ filestream-monitoring
   │  │  ├─ status: (HEALTHY) Healthy
   │  │  └─ type: OUTPUT
   │  └─ filestream-monitoring-filestream-monitoring-agent
   │     ├─ status: (HEALTHY) Healthy
   │     └─ type: INPUT
   ├─ http/metrics-monitoring
   │  ├─ status: (HEALTHY) Healthy: communicating with pid '75188'
   │  ├─ http/metrics-monitoring
   │  │  ├─ status: (HEALTHY) Healthy
   │  │  └─ type: OUTPUT
   │  └─ http/metrics-monitoring-metrics-monitoring-agent
   │     ├─ status: (HEALTHY) Healthy
   │     └─ type: INPUT
   └─ system/metrics-default
      ├─ status: (HEALTHY) Healthy: communicating with pid '75185'
      ├─ system/metrics-default
      │  ├─ status: (HEALTHY) Healthy
      │  └─ type: OUTPUT
      └─ system/metrics-default-unique-system-metrics-input
         ├─ status: (HEALTHY) Healthy
         └─ type: INPUT
~/Downloads/builds/elastic-agent-8.14.0-SNAPSHOT-darwin-aarch64 ··························· 08:38:38 PM
❯ sudo ls /Library/Elastic/Agent/data/tmp/
Hk6rvk9TDibMPcDvpl0jkLE-qDsHWVYL.sock   iThI_df0cBKC6YUNGGlKscMkOfz3FBH3.sock
akSPbdqgaHaTY0_J01-dsfYK6JpMz2zn.sock
```

**Comment by cmacknz** [comment]
```
❯ ps aux | rg beat
cmackenzie       79016   0.0  0.0 410066224     48 s052  S+    8:40PM   0:00.00 rg beat
root             75188   0.0  0.4 412278336 145824   ??  S     8:36PM   0:00.41 /Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/components/agentbeat metricbeat -E setup.ilm.enabled=false -E setup.template.enabled=false -E management.enabled=true -E management.restart_on_output_change=true -E logging.level=info -E logging.to_stderr=true -E gc_percent=${METRICBEAT_GOGC:100} -E metricbeat.config.modules.enabled=false -E http.enabled=true -E http.host=unix:///Library/Elastic/Agent/data/tmp/akSPbdqgaHaTY0_J01-dsfYK6JpMz2zn.sock -E path.data=/Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/run/http/metrics-monitoring
root             75187   0.0  0.4 412295936 147472   ??  S     8:36PM   0:00.44 /Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/components/agentbeat metricbeat -E setup.ilm.enabled=false -E setup.template.enabled=false -E management.enabled=true -E management.restart_on_output_change=true -E logging.level=info -E logging.to_stderr=true -E gc_percent=${METRICBEAT_GOGC:100} -E metricbeat.config.modules.enabled=false -E http.enabled=true -E http.host=unix:///Library/Elastic/Agent/data/tmp/Hk6rvk9TDibMPcDvpl0jkLE-qDsHWVYL.sock -E path.data=/Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/run/beat/metrics-monitoring
root             75186   0.0  0.4 412243552 141776   ??  S     8:36PM   0:00.54 /Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/components/agentbeat filebeat -E setup.ilm.enabled=false -E setup.template.enabled=false -E management.enabled=true -E management.restart_on_output_change=true -E logging.level=info -E logging.to_stderr=true -E gc_percent=${FILEBEAT_GOGC:100} -E filebeat.config.modules.enabled=false -E path.data=/Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/run/filestream-monitoring
root             75185   0.0  0.4 412237216 149696   ??  S     8:36PM   0:00.98 /Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/components/agentbeat metricbeat -E setup.ilm.enabled=false -E setup.template.enabled=false -E management.enabled=true -E management.restart_on_output_change=true -E logging.level=info -E logging.to_stderr=true -E gc_percent=${METRICBEAT_GOGC:100} -E metricbeat.config.modules.enabled=false -E http.enabled=true -E http.host=unix:///Library/Elastic/Agent/data/tmp/iThI_df0cBKC6YUNGGlKscMkOfz3FBH3.sock -E path.data=/Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/run/system/metrics-default
```

The filestream monitoring isn't getting a monitoring unix socket, and it appears as a component in the output of `elastic-agent status`, which is causing the leak detection test to fail:

```
root             75186   0.0  0.4 412243552 141776   ??  S     8:36PM   0:00.54 /Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/components/agentbeat filebeat -E setup.ilm.enabled=false -E setup.template.enabled=false -E management.enabled=true -E management.restart_on_output_change=true -E logging.level=info -E logging.to_stderr=true -E gc_percent=${FILEBEAT_GOGC:100} -E filebeat.config.modules.enabled=false -E path.data=/Library/Elastic/Agent/data/elastic-agent-8.14.0-SNAPSHOT-d44fe9/run/filestream-monitoring
```

**Comment by cmacknz** [comment]
I still see attempts to index to the non-existant `elastic_agent.agentbeat` datastream:

```json
{"log.level":"warn","@timestamp":"2024-04-18T00:44:26.369Z","message":"Cannot index event (status=403): dropping event! Enable debug logs to view the event and cause.","component":{"binary":"agentbeat","dataset":"elastic_agent.agentbeat","id":"filestream-monitoring","type":"filestream"},"log":{"source":"filestream-monitoring"},"service.name":"filebeat","ecs.version":"1.6.0","log.logger":"elasticsearch","log.origin":{"file.line":429,"file.name":"elasticsearch/client.go","function":"github.com/elastic/beats/v7/libbeat/outputs/elasticsearch.(*Client).bulkCollectPublishFails"},"ecs.version":"1.6.0"}
```

The list of data streams for monitoring is hard coded into Fleet, we need a Kibana change for this to be accepted. Regardless of the permission issue, we need to use the correct beat alias (filebeat, metricbeat, etc) to avoid breaking all of the elastic_agent dashboards.

**Comment by cmacknz** [comment]
The logs ingestion test should be failing as well giving us a stronger hint to the problem, but it only looks for metricbeat logs: https://github.com/elastic/elastic-agent/blob/d44fe9fac2a56d529ddf30833fd2b70535588249/testing/integration/logs_ingestion_test.go#L246-L253

The unix socket connection errors are also in the agent logs:

```json
{"log.level":"error","@timestamp":"2024-04-18T00:50:06.000Z","message":"Error fetching data for metricset beat.stats: error making http request: Get \"http://unix/stats\": dial unix /Library/Elastic/Agent/data/tmp/xTEtpJ7117ppc6OYvJCaYHbDW8mLjXGe.sock: connect: no such file or directory","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"log":{"source":"beat/metrics-monitoring"},"ecs.version":"1.6.0","log.origin":{"file.line":256,"file.name":"module/wrapper.go","function":"github.com/elastic/beats/v7/metricbeat/mb/module.(*metricSetWrapper).fetch"},"service.name":"metricbeat","ecs.version":"1.6.0"}
```

The error detection logic should also be failing here but I don't think it waits long enough to see all possible errors: https://github.com/blakerouse/elastic-agent/blob/d44fe9fac2a56d529ddf30833fd2b70535588249/testing/integration/logs_ingestion_test.go#L266-L269

**Comment by blakerouse** [comment]
@cmacknz The issue is actually in the `agentbeat.spec.yml` that is defined in the beats repository.

https://github.com/elastic/beats/issues/39025

**Comment by blakerouse** [comment]
PR to fix the specification - https://github.com/elastic/beats/pull/39026

**Comment by pchila** [comment]
buildkite test this

**Comment by pierrehilbert** [comment]
After https://github.com/elastic/beats/pull/39026 has been merged, @pchila ran tests locally and the failure is fixed. 
As discussed with @jlind23, to avoid blocking the first BC, I'm force merging this one.

**Pull request merged**
